File: Java/Valid-Parentheses-Easy.java
// Problem Link: https://leetcode.com/problems/valid-parentheses/description/

// Space Complexity: O(n)
//  Time Complexity: O(n)

// Performace:
//     RunTime:   2 ms      ( better then 54,18% of users of Java )
//      Memory:   41.59 MB  ( better then 32,21% of users of Java ) 

package Java;
import java.util.ArrayList;

class Solution {
  public boolean isValid(String s) {
    if (s.length() % 2 != 0) return false;

    ArrayList<Character> openLeft = new ArrayList<Character>();

    String open = "([{";

    for (int i = 0; i < s.length(); i++) {
      if (open.contains(String.valueOf(s.charAt(i)))) {
        openLeft.add(s.charAt(i));
      } else {
        if(openLeft.size() == 0) return false;
        if (openLeft.get(openLeft.size()-1) == returnThePair(s.charAt(i))) {
          openLeft.remove(openLeft.size()-1);
        } else {
            return false;
        }
      }
    }

    if (openLeft.isEmpty()) return true;
    return false;
  }

  public Character returnThePair(Character c) {
    switch (c) {
      case ')':
        return '(';
      case ']':
        return '[';
      default:
        return '{';
    }
  }
}

File: JavaScript/Valid-Parentheses-Easy.js
// Problem Link: https://leetcode.com/problems/valid-parentheses/description/

// Space Complexity: O(n) 
//  Time Complexity: O(n)

// Performace:
//     RunTime:   56    ms  ( better then 65,01% of users of JavaScript )
//      Memory:   50.08 MB  ( better then 61,64% of users of JavaScript ) 

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) { 
  let open = "([{"
  let close = ")]}"

  let openStack = []
  for(let i=0; i<s.length; i++) {
    if (open.includes(s.charAt(i))) {
      openStack.push(s.charAt(i))
    } else {
      if (openStack[openStack.length-1] === open[close.indexOf(s.charAt(i))]) {
        openStack.pop()
      } else {
        return false
      }
    }
  }
  if (openStack.length === 0) return true

  return false
};